tfx/extensions/google_cloud_kubernetes/runner.py

```diff
@@ -0,0 +1,245 @@
+_TFX_IMAGE = 'tensorflow/tfx:%s' % (version.__version__)
+
+_COMMAND = ["python", "-m", "tfx.scripts.run_executor"]
+
+
+def _build_pod_names(num_workers: int, unique_id: Text) -> List[Text]:
+  return ['training-worker-{}-{}'.format(unique_id,
+                                         i) for i in range(num_workers)]
+
+
+def _build_service_names(num_workers: int, unique_id: Text) -> List[Text]:
+  return ['training-service-{}-{}'.format(unique_id,
+                                          i) for i in range(num_workers)]
+
+
+def _pod_is_done(resp: client.V1Pod):
+  return kube_utils.PodPhase(resp.status.phase).is_done
+
+
+def create_worker_pods(job_args: List[Text],
+                       training_inputs: Dict[Text, Any],
+                       unique_id: Text):
+  """Create worker pods for multi-worker training."""
+  tfx_image = training_inputs.get('tfx_image', _TFX_IMAGE)
+  num_workers = training_inputs.get('num_workers', 1)
+  num_gpus_per_worker = training_inputs.get('num_gpus_per_worker', 0)
+
+  api_instance = kube_utils.make_core_v1_api()
+  service_names = _build_service_names(num_workers=num_workers,
+                                       unique_id=unique_id)
+  pod_names = _build_pod_names(num_workers=num_workers, unique_id=unique_id)
+  worker_hosts = ['{}:5000'.format(svc_name) for svc_name in service_names]
+
+  # TODO(ericlege): consider using a jinja2 template instead
+  for i in range(num_workers):
+    tf_config = json.dumps({
+        'cluster': {
+            'worker': worker_hosts
+        },
+        'task': {'type': 'worker', 'index': i}
+    })
+    pod = client.V1Pod(
+        metadata=client.V1ObjectMeta(
+            name=pod_names[i],
+            labels={
+                'name': 'training',
+                'id': unique_id,
+                'task': str(i),
+            },
+        ),
+        spec=client.V1PodSpec(
+            containers=[
+                client.V1Container(
+                    name='worker-pod',
+                    image=tfx_image,
+                    command=_COMMAND,
+                    args=job_args,
+                    security_context=client.V1SecurityContext(
+                        privileged=True,
+                    ),
+                    env=[
+                        client.V1EnvVar(
+                            name='TF_CONFIG',
+                            value=tf_config,
+                        ),
+                    ],
+                    ports=[
+                        client.V1ContainerPort(
+                            container_port=5000,
+                        ),
+                    ],
+                    resources=client.V1ResourceRequirements(
+                        limits={
+                            'nvidia.com/gpu': num_gpus_per_worker,
+                        },
+                    ) if num_gpus_per_worker > 0 else None,
+                ),
+            ],
+            restart_policy=kube_utils.RestartPolicy.NEVER.value,
```

Comment:
What's the rationale for not allowing restarts?

Reply:
I think Kubernetes by default restarts completed containers, so in the event of successful training we would like to prevent the pod from running twice; if the pod fails, then we also do not need to restart it since the component would also fail as well.